### PR TITLE
FYLE-2a81g3v: Moved tax group service to platform and added tax group models

### PR DESCRIPTION
**Pull request by Dimple16**

### CODE REVIEW

File: src/app/core/models/platform/platform-tax-group.model.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { PlatformTaxGroupData } from '../../models/platform/platform-tax-group-data.model';
+
+export interface PlatformTaxGroup {
+  count: number;
+  offset: number;
+  data: PlatformTaxGroupData[];
```
Comment (rvab) [inline src/app/core/models/platform/platform-tax-group.model.ts:6]: Make this a generic interface.
Screenshot 2022-05-19 at 8 55 44 PM
Screenshot 2022-05-19 at 8 56 02 PM

File: src/app/core/services/tax_group.service.ts [hunk 2]
```diff
@@ -1,21 +1,67 @@
...
+      },
+    };
+    return this.spenderPlatformApiService
+      .get('/tax_groups', data)
+      .pipe(map((res: PlatformTaxGroup) => this.transformFrom(res.data)));
+  }
+
+  transformFrom(platformTaxGroup: PlatformTaxGroupData[]): TaxGroup[] {
+    let oldTaxGroup = [];
+    oldTaxGroup = platformTaxGroup.map((taxGroup) => ({
```
Comment (rvab) [inline src/app/core/services/tax_group.service.ts:42]: use const here. No need of let variable?

File: src/app/core/services/tax_group.service.ts [hunk 3]
```diff
@@ -1,21 +1,67 @@
...
+      id: taxGroup.id,
+      name: taxGroup.name,
+      percentage: taxGroup.percentage,
+      created_at: taxGroup.created_at,
+      updated_at: taxGroup.updated_at,
+      org_id: taxGroup.org_id,
+      is_enabled: taxGroup.is_enabled,
+    }));
```
Comment (rvab) [inline src/app/core/services/tax_group.service.ts:50]: Why do we need to transform it into an old Interface?. Can't we use the new Interface?

File: src/app/core/services/tax_group.service.ts [hunk 4]
```diff
@@ -1,21 +1,67 @@
...
-    return this.apiV2Service.get('/tax_groups', data).pipe(map((res) => res.data));
+  get(): Observable<TaxGroup[]> {
+    return this.getEnabledTaxGroupsCount().pipe(
+      switchMap((count) => {
+        count = count > 50 ? count / 50 : 1;
+        return range(0, count);
+      }),
+      concatMap((page) => this.getTaxGroups({ offset: 50 * page, limit: 50 })),
+      map((res) => res),
+      reduce((acc, curr) => acc.concat(curr), [] as any[])
```
Comment (rvab) [inline src/app/core/services/tax_group.service.ts:63]: Can we avoid `any`?

File: src/app/core/services/tax_group.service.ts [hunk 5]
```diff
@@ -1,21 +1,67 @@
 import { Injectable } from '@angular/core';
```
Comment (rvab) [inline src/app/core/services/tax_group.service.ts:1]: Shouldn't the file name be `tax-group.service.ts`?

File: src/app/core/services/tax_group.service.ts [hunk 6]
```diff
@@ -1,21 +1,67 @@
...
 import { ApiService } from './api.service';
-import { map } from 'rxjs/operators';
-import { ApiV2Service } from './api-v2.service';
+import { concatMap, map, reduce, switchMap } from 'rxjs/operators';
 import { TaxGroup } from '../models/tax_group.model';
```
Comment (rvab) [inline src/app/core/services/tax_group.service.ts:4]: Shouldn't the file name be `tax-group.model`?

**Review by rvab — CHANGES_REQUESTED** [review]

**Review by rvab — CHANGES_REQUESTED** [review]
Please take care of the comments

File: src/app/core/services/tax-group.service.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import { Injectable } from '@angular/core';
```
Comment (Dimple16) [inline src/app/core/services/tax-group.service.ts:1]: renamed service filed

File: src/app/core/services/tax_group.service.ts (hunk 3 above)
Comment (Dimple16) [reply src/app/core/services/tax_group.service.ts]: we decided to move the older APIs to the platform in this way,
we'll be using transform methods in service to transform the data,
this way, we can avoid making changes everywhere used,

File: src/app/core/services/tax-group.service.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        limit: config.limit,
+      },
+    };
+    return this.spenderPlatformApiService
+      .get<PlatformApiResponse<PlatformTaxGroup>>('/tax_groups', data)
+      .pipe(map((res) => this.transformFrom(res.data)));
+  }
+
+  transformFrom(platformTaxGroup: PlatformTaxGroup[]): TaxGroup[] {
+    const oldTaxGroup = platformTaxGroup.map((taxGroup) => ({
```
Comment (rvab) [inline src/app/core/services/tax-group.service.ts:43]: Can we take up the transformation in the next iteration?. We should make this change

File: src/app/core/services/tax-group.service.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        count = count > 50 ? count / 50 : 1;
+        return range(0, count);
+      }),
+      concatMap((page) => this.getTaxGroups({ offset: 50 * page, limit: 50 })),
+      map((res) => res),
+      reduce((acc, curr) => acc.concat(curr), [] as TaxGroup[])
+    );
+  }
+
+  post(taxGroup: TaxGroup) {
```
Comment (rvab) [inline src/app/core/services/tax-group.service.ts:68]: return type?

**Review by rvab — APPROVED** [review]
Minor comments

File: src/app/core/services/tax-group.service.ts (hunk 8 above)
Comment (Dimple16) [reply src/app/core/services/tax-group.service.ts:43]: we decided to do the transformation for all APIs the same way

File: src/app/core/services/tax-group.service.ts (hunk 9 above)
Comment (Dimple16) [reply src/app/core/services/tax-group.service.ts:68]: haven't made changes here, this is old code, will take care of this in post call changes pr

File: src/app/core/services/tax-group.service.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class TaxGroupService {
+  constructor(private apiService: ApiService, private spenderPlatformApiService: SpenderPlatformApiService) {}
+
+  getEnabledTaxGroupsCount(): Observable<number> {
+    const data = {
+      params: {
+        is_enabled: 'eq.' + true,
```
Comment (abhishek1234321) [inline src/app/core/services/tax-group.service.ts:19]: is this needed here? I saw you removed it somewhere, doesn't platform give enabled only by default?

File: src/app/core/services/tax-group.service.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    };
+    return this.spenderPlatformApiService
+      .get<PlatformApiResponse<PlatformTaxGroup>>('/tax_groups', data)
+      .pipe(map((res) => res.count));
+  }
+
+  getTaxGroups(config: { offset: number; limit: number }): Observable<TaxGroup[]> {
+    const data = {
+      params: {
+        is_enabled: 'eq.' + true,
```
Comment (abhishek1234321) [inline src/app/core/services/tax-group.service.ts:32]: same as above

**Review by abhishek1234321 — CHANGES_REQUESTED** [review]

File: src/app/core/services/tax-group.service.ts (hunk 10 above)
Comment (Dimple16) [reply src/app/core/services/tax-group.service.ts:19]: yes, I removed it from the offline service and added it here like other services

**Review by abhishek1234321 — APPROVED** [review]

**Pull request merged**